pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -1064,3 +1064,65 @@ func setUpstreamProtocol(node *model.Proxy, c *cluster.Cluster, port *model.Port
 		c.ProtocolSelection = cluster.Cluster_USE_DOWNSTREAM_PROTOCOL
 	}
 }
+
+func addTelemetryMetadata(opts buildClusterOpts) {
+	if opts.cluster == nil || opts.proxy == nil || opts.proxy.ServiceInstances == nil ||
+		len(opts.proxy.ServiceInstances) == 0 {
+		return
+	}
+	if opts.cluster.Metadata == nil {
+		opts.cluster.Metadata = &core.Metadata{
+			FilterMetadata: map[string]*structpb.Struct{},
+		}
+	}
+
+	// Create Istio metadata if does not exist yet
+	if _, ok := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]; !ok {
+		opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey] = &structpb.Struct{
+			Fields: map[string]*structpb.Value{},
+		}
+	}
+
+	im := opts.cluster.Metadata.FilterMetadata[util.IstioMetadataKey]
+
+	// If services already exist in cluster metadata, delete it and recreate the field,
+	// since service metadata should only be added only in this function.
+	delete(im.Fields, "services")
+
+	// Add services field into istio metadata
+	im.Fields["services"] = &structpb.Value{
+		Kind: &structpb.Value_ListValue{
+			ListValue: &structpb.ListValue{
+				Values: []*structpb.Value{},
+			},
+		},
+	}
+
+	svcMetaList := im.Fields["services"].GetListValue()
+	// Add servie related metadata. This will be consumed by telemetry v2 filter for metric labels.
+	for _, svc := range opts.proxy.ServiceInstances {
```

Comment:
I am not sure serviceInstances is right here. This would include multiple duplicate entries in some cases, as well as the wrong service. For example, if i have svcFoo with port 80 and svcBar with port 90, ServiceInstance will have both so port 80 cluster will have foo and bar. It should just have foo.

Reply:
Yeah and also this is not the right source for outbound cluster label. Updated the function signature to take Service and traffic direction. At inbound, port will be used to distinguish services that will be added to the cluster metadata.